test/lsp/multithreaded_protocol_test_corpus.cc

```diff
@@ -19,6 +27,99 @@ TEST_P(ProtocolTest, MultithreadedWrapperWorks) {
     assertDiagnostics(send(move(requests)), {{"yolo1.rb", 3, "bear"}});
 }
 
+TEST_P(ProtocolTest, CancelsSlowPathWhenNewEditWouldTakeFastPathWithOldEdits) {
+    auto initOptions = make_unique<SorbetInitializationOptions>();
+    initOptions->enableTypecheckInfo = true;
+    assertDiagnostics(initializeLSP(true /* supportsMarkdown */, move(initOptions)), {});
+
+    // Create three files.
+    assertDiagnostics(send(*openFile("foo.rb", "# typed: true\n\nclass Foo\n\nend\n")), {});
+    assertDiagnostics(
+        send(*openFile(
+            "bar.rb",
+            "# typed: true\n\nclass Bar\nextend T::Sig\n\nsig{returns(String)}\ndef hello\n\"hi\"\nend\nend\n")),
+        {});
+    // baz calls the method defined in bar
+    assertDiagnostics(send(*openFile("baz.rb", "# typed: true\n\nclass Baz\nextend "
+                                               "T::Sig\n\nsig{returns(String)}\ndef hello\nBar.new.hello\nend\nend\n")),
+                      {});
+
+    // Slow path edits two files. One introduces error.
+    sendAsync(LSPMessage(make_unique<NotificationMessage>("2.0", LSPMethod::PAUSE, nullopt)));
+    // Syntax error in foo.rb.
+    sendAsync(*changeFile("foo.rb", "# typed: true\n\nclass Foo\ndef noend\nend\n", 2, true));
+    // Typechecking error in bar.rb
+    sendAsync(*changeFile(
+        "bar.rb", "# typed: true\n\nclass Bar\nextend T::Sig\n\nsig{returns(Integer)}\ndef hello\n\"hi\"\nend\nend\n",
+        2, true));
+    sendAsync(LSPMessage(make_unique<NotificationMessage>("2.0", LSPMethod::RESUME, nullopt)));
+    // Wait for typechecking to begin to avoid races.
+    {
+        auto status = getTypecheckRunStatus(*readAsync());
+        ASSERT_TRUE(status.has_value());
+        ASSERT_EQ(*status, SorbetTypecheckRunStatus::Started);
+    }
+
+    // Make another edit that fixes syntax error and should take fast path.
+    sendAsync(*changeFile("foo.rb", "# typed: true\n\nclass Foo\nend\n", 2, false));
```

Comment:
Observation: this reverts `foo.rb` to the state of the files at line 36.  Does this take the fast path because it reverted to a state that has already been previously parsed, or does it work regardless of previous parsings because it fixed the syntax error?

(The comment on line 63 made me think that it's the latter, in which case it would have been slightly clearer to me if line 64 added the `end` to `def noend`, rather than removing the method declaration.)

Reply:
It takes the fast path because it fixes the syntax error and doesn't make any structural changes to the code (e.g., new methods).